World builder structure

MFraters opened the pull request:
@bangerth is this in the direction we discussed?

I also looked into translating the input into json, and I have currently two different options of which I think both would do, so that is the direction I will try out.

MFraters commented:
Btw, one of the downsides of the boost json parser is that is stores all values as strings, and therefore can't parse the types. There are way more advanced json libraries out there, but I think I can also just parse it myself (which is the option I am now going to try).

bangerth commented on include/aspect/world_builder/interface.h:
```diff
@@ -0,0 +1,214 @@
+     * @ingroup WorldBuilder
+     */
+    template <int dim>
+    class Interface //: public SimulatorAccess<dim>
```
It's not clear to me what this is actually an interface for. A description of a part of the earth? Maybe move the interface class into a (sub-)namespace with a name that makes this clear.

MFraters replied:
It is an interface for the world builder plugins. I expended the description.

bangerth commented on source/simulator/core.cc:
```diff
@@ -152,6 +153,7 @@ namespace aspect
     parameters (prm, mpi_communicator_),
     melt_handler (parameters.include_melt_transport ? new MeltHandler<dim> (prm) : NULL),
     newton_handler (parameters.nonlinear_solver == NonlinearSolver::Newton_Stokes ? new NewtonHandler<dim> () : NULL),
+    world_builder (parameters.world_builder_directory != "" ? WorldBuilder::create_world_builder<dim> (prm) : NULL),
```
I would always create such an object. It may simply not be used if no plugin references it.

The presence of the `world_builder_directory` variable in the main class confuses me. What is it used for? Why is that not a parameter of the world builder itself, and located in the corresponding structure?

MFraters replied:
>I would always create such an object. It may simply not be used if no plugin references it.

sure.

>The presence of the `world_builder_directory` variable in the main class confuses me. What is it used for? Why is that not a parameter of the world builder itself, and located in the corresponding structure?

It should have been `world_builder_file`. But now I won't need it now we always create the world builder, so I removed it

bangerth commented on source/simulator/parameters.cc:
```diff
@@ -245,6 +246,10 @@ namespace aspect
 
+    prm.declare_entry ("World builder file", "",
+                       Patterns::FileName (),
+                       "The location of the world builder json file.");
```
JSON

bangerth commented on source/simulator/parameters.cc:
```diff
@@ -1055,6 +1060,9 @@ namespace aspect
                                  mpi_communicator,
                                  false);
 
+    world_builder_directory = prm.get("World builder file");
```
is it a file or a directory? ;-)

bangerth commented on source/world_builder/interface.cc:
```diff
@@ -0,0 +1,227 @@
+      }
+      //prm.leave_subsection ();
+
+      sort(model_names.begin(), model_names.end());
```
std::sort

bangerth commented on source/world_builder/interface.cc:
```diff
@@ -0,0 +1,227 @@
+
+      sort(model_names.begin(), model_names.end());
+      std::vector<std::string>::iterator it;
+      it = unique(model_names.begin(), model_names.end());
```
std::unique

MFraters commented:
I have now build it into a version which can actually do what I want it to do and addressed the comments. Because the parameter handler of deal.ii is currently not capable to what I need to properly read in JSON (made a start on it in https://github.com/dealii/dealii/pull/5993), but even if I would get everything in into deal.ii 9.0 before it is released, it won't be in deal.ii 8.5. So I decided for now to focus on implementing it here in aspect locally in a way that is designed to be as compatible as possible with the deal.ii parameter handler and the internal structure of the boost property tree in it. 

I added an example world builder file (wb3.json) and a empty plugin which declares values to show that the values are correctly declared and parsed.

gassmoeller commented on build_9.0/wb3.json:
```diff
@@ -0,0 +1,18 @@
+{
```
We will probably need a new directory in `data/world-builder/` or `data/initial-temperature/world-builder` for these files.
Also: Is it possible to add comments to these files?

MFraters replied:
>We will probably need a new directory in `data/world-builder/` or `data/initial-temperature/world-builder` for these files.

I think it will be better to put it in the end into a cookbook, since it is not really data, just an example input.

>Also: Is it possible to add comments to these files?

Unfortunately for us, they removed support for comments from JSON because people where apparently abusing it as macro's, and some other reasons (see https://groups.yahoo.com/neo/groups/json/conversations/topics/156 ).  But what can be done is just adding a field `"comment":"and put the comment here",`. I could them exempt the comment field from needing declaration, or ignore the field when reading stuff in. 

An other thing I could manually parse the file in advance and remove comment lines, but then we would diverge from the JSON standard, making implementation into other codes slightly more difficult.

bangerth replied:
Yes, manually parsing is not an option and extending the file format with comments is also not an option. I do like the idea of a "comment" field.

gassmoeller replied:
Ok, that is good enough for me. I just want to have the option to put in stuff like "this is model X with data taken from publication Y" or something of that sort.

gassmoeller commented on include/aspect/parameters.h:
```diff
@@ -319,6 +319,7 @@ namespace aspect
     bool                           use_conduction_timestep;
     bool                           convert_to_years;
     std::string                    output_directory;
+    std::string                    world_builder_directory;
```
Hmm, why do you need this variable? Could it not be a member of the world builder?

gassmoeller commented on include/aspect/simulator_access.h:
```diff
@@ -663,6 +668,13 @@ namespace aspect
       get_newton_handler () const;
 
+      /**
+       * Return a reference to the world builder handler that controls the
```
I would just call it world builder, the 'handler' is unnecessary.

gassmoeller commented on include/aspect/world_builder/continental_plate.h:
```diff
@@ -0,0 +1,83 @@
+/*
+  Copyright (C) 2011 - 2016 by the authors of the ASPECT code.
```
2018 (or whenever you started it until 2018)

gassmoeller commented on include/aspect/world_builder/continental_plate.h:
```diff
@@ -0,0 +1,83 @@
+*/
+
+
+#ifndef _aspect_initial_temperature_world_generator_modules_constant_h
```
world_builder_continental_plate_h
the same in the line below

gassmoeller commented on include/aspect/world_builder/continental_plate.h:
```diff
@@ -0,0 +1,83 @@
+    using namespace dealii;
+
+    /**
+     * A class that implements zero initial topography.
```
Needs a new comment

gassmoeller commented on include/aspect/world_builder/continental_plate.h:
```diff
@@ -0,0 +1,83 @@
+     * @ingroup InitialTopographyModels
+     */
+    template <int dim>
+    class Constant : public Interface<dim>
```
Needs a new class name?

gassmoeller commented on include/aspect/world_builder/continental_plate.h:
```diff
@@ -0,0 +1,83 @@
+        virtual
+        void temperature (const Point<dim> &position,
+                          const unsigned int object_number,
+                          const WorldBuilderParameters<dim> &world_generator_parameters,
```
Hmm, I do not like this very much. Which `WorldBuilderParameters` might be needed by this plugin? Should not all things either be World builder parameter (in which case World Builder handles whatever should happen with them), or local parameters of this plugin?

gassmoeller replied:
I personally prefer to hand over the previous temperature as `const double temperature` and return the new temperature as `double`. But that might be a personal preference.

MFraters replied:
Yes, this had to do with the old structure, I still have to rethink the approach. I only added this file for now to show that the plugin structure with the declaration works.

gassmoeller commented on source/world_builder/interface.cc:
```diff
@@ -0,0 +1,681 @@
+/*
+  Copyright (C) 2011 - 2016 by the authors of the ASPECT code.
```
2018

gassmoeller commented on source/world_builder/interface.cc:
```diff
@@ -0,0 +1,681 @@
+
+
+#include <aspect/global.h>
+//#include <aspect/world_builder.h>
```
Remove

gassmoeller commented on include/aspect/world_builder/interface.h:
```diff
@@ -0,0 +1,269 @@
+     */
+    template <int dim>
+    void
+    register_world_builder (const std::string &name,
```
This function and the `create_world_builder` function are not the best way to implement this. I suppose you took this from one of the other plugin systems without manager, but it is simpler to use the same structure as for the `InitialTemperature::Manager` class. All the stuff in `create_world_builder` can either move into the constructor of `Manager` (if it is related to the core of `Manager`), or into `parse_parameters` (if it is related to the plugins). `register_world_builder` can become a member function of `Manager`, and `declare_parameters` can also become member functions of `Manager`.

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -152,6 +153,7 @@ namespace aspect
     parameters (prm, mpi_communicator_),
     melt_handler (parameters.include_melt_transport ? new MeltHandler<dim> (prm) : NULL),
     newton_handler (parameters.nonlinear_solver == NonlinearSolver::Newton_Stokes ? new NewtonHandler<dim> () : NULL),
+    world_builder (WorldBuilder::create_world_builder<dim> (prm)),
```
If you move the functions as I mentioned above, you can change this to the same structure as for the NewtonHandler or MeltHandler.

MFraters commented:
Thanks for the comments! I have addressed them and implemented the manager temperature and composition function. I still have to do the parsing, and then I think I can get this out of the WIP phase.

MFraters commented:
The code is now ready for review. I will add a test tomorrow.

tjhei commented on include/aspect/world_builder/interface.h:
```diff
@@ -0,0 +1,316 @@
+    };
+
+
+    using namespace dealii;
```
redundant, see above.

tjhei commented on include/aspect/world_builder/interface.h:
```diff
@@ -0,0 +1,316 @@
+    {
+      public:
+        WorldBuilderParameterHandler(const char path_separator = '.');
+        void set_tree(boost::property_tree::ptree &tree);
```
please document these functions. I have no idea what they do.

tjhei commented on include/aspect/world_builder/interface.h:
```diff
@@ -0,0 +1,316 @@
+
+    /**
+     * Base class for classes that describe particular initial conditions.
+     * All plugins, which describe a plate tectonic features in terms of
```
grammar, remove "a"?

tjhei commented on include/aspect/world_builder/interface.h:
```diff
@@ -0,0 +1,316 @@
+
+
+
+      private:
```
remove

tjhei commented on include/aspect/world_builder/interface.h:
```diff
@@ -0,0 +1,316 @@
+        static void
+        declare_parameters (ParameterHandler &prm);
+
+        ParameterHandler wb_prm;
```
are these supposed to be public?

tjhei commented on include/aspect/world_builder/interface.h:
```diff
@@ -0,0 +1,316 @@
+        double potential_mantle_temperature, thermal_expansion_coefficient_alfa, specific_heat_Cp;
+
+        //const std_cxx11::shared_ptr<GravityModel::Interface<dim> > gravity_model;
+        //const std_cxx11::shared_ptr<GeometryModel::Interface<dim> > geometry_model;
```
remove

tjhei commented on source/utilities.cc:
```diff
@@ -2468,6 +2470,49 @@ namespace aspect
     }
 
+    /**
+     * todo
```
remove the comment here ;-)

tjhei commented on source/utilities.cc:
```diff
@@ -2468,6 +2470,49 @@ namespace aspect
+     */
+    template <int dim>
+    std::vector<Point<dim> >
+    vector_double_to_point(const std::vector<std::vector<double> > &s)
```
these functions would make a good separate PR, don't you think?

tjhei commented on source/utilities.cc:
```diff
@@ -2573,7 +2618,7 @@ namespace aspect
 
     template <int dim>
-    std_cxx11::array<double,dim> &NaturalCoordinate<dim>::get_coordinates()
+    const std_cxx11::array<double,dim> NaturalCoordinate<dim>::get_coordinates() const
```
also good as a separate PR!

tjhei commented on source/world_builder/interface.cc:
```diff
@@ -0,0 +1,1082 @@
+        u.reserve (s.size());
+
+        // see if the name is special and if so mangle the whole thing
+        const bool mangle_whole_string = (s == "value");
```
I don't understand this, if s==value, couldn't you just return the mangled name of "value"?

tjhei commented on source/world_builder/interface.cc:
```diff
@@ -0,0 +1,1082 @@
+              switch (s[i + 1])
+                {
+                  case '0':
+                    c = 0 * 16;
```
you can use something like ``(s[i+1]-'0')`` to get the value if it is a digit.

tjhei commented on source/world_builder/interface.cc:
```diff
@@ -0,0 +1,1082 @@
+                  default:
+                    Assert(false, ExcInternalError());
+                }
+              switch (s[i + 2])
```
you can use a for loop to handle both chars with a single logic (and multiply by 16 in between)

tjhei commented on source/world_builder/interface.cc:
```diff
@@ -0,0 +1,1082 @@
+      double temperature;
+      if (dim==2)
+        {
+          //Todo, convert from a 2d point of a cross section to a 3d point for the world generator
```
put an assert in here.
